### PR TITLE
feat(postgres): added postgres support

### PR DESCRIPTION
**Pull request by DmitriyLewen**

### CODE REVIEW

File: dbservice/postgresdb/dbparam.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package postgresdb
+
+import (
+	"errors"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+)
+
+var (
```
Comment (AndreyLevchenko) [inline dbservice/postgresdb/dbparam.go:10]: the vars are not specific to postgres. They could be stored in some common package

File: dbservice/postgresdb/init.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+)
+
+var (
+	tableSchemas = map[string]string{
+		dbTableName:         "CREATE TABLE IF NOT EXISTS %s (id text, messagekey text,messagevalue text);",
```
[the remark this answers is not in the document]
Comment (AndreyLevchenko) [reply dbservice/postgresdb/init.go:11]: Another question: bolt accepts `bytes`. Can we use bytes with postgres?

File: dbservice/postgresdb/insert.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+)
+
+var insert = func(db *sqlx.DB, table, id, columnName2, value2, columnName3, value3 string) error {
+	var i int
+	if err := db.Get(&i, fmt.Sprintf("SELECT COUNT(*) FROM %s WHERE (%s=$1 AND %s=$2)", table, "id", columnName2), id, value2); err != nil {
```
Comment (AndreyLevchenko) [inline dbservice/postgresdb/insert.go:11]: why you are using constant `"id"` as parameter to `fmt.Sprintf()`?

File: msgservice/applicationscopeowner_test.go [hunk 4]
```diff
@@ -21,12 +22,16 @@ var (
 )
 
 func TestApplicationScopeOwner(t *testing.T) {
-	dbPathReal := dbservice.DbPath
+	db = boltdb.NewBoltDb()
+	oldDb := dbservice.Db
+	dbservice.Db = db
+	defer func() { dbservice.Db = oldDb }()
```
Comment (AndreyLevchenko) [inline msgservice/applicationscopeowner_test.go:28]: use single `defer`

**Review by AndreyLevchenko — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -340,7 +340,7 @@ The Postee container uses BoltDB to store information about previously scanned i
 This is used to prevent resending messages that were already sent before.
 The size of the database can grow over time. Every image that is saved in the database uses 20K of storage.  
  
-Postee supports ‘PATH_TO_DB’ environment variable to change the database directory. To use, set the ‘PATH_TO_DB’ environment variable to point to the database file, for example: PATH_TO_DB="./database/webhook.db". By default, the directory for the database file is “/server/database/webhook.db”.
+Postee supports ‘PATH_TO_BOLTDB’ environment variable to change the bolt database directory. To use, set the ‘PATH_TO_BOLTDB’ environment variable to point to the bolt database file, for example: PATH_TO_BOLTDB="./database/webhook.db". By default, the directory for the bolt database file is “/server/database/webhook.db”.
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply README.md]: done

File: dbservice/boltdb/checker.go [hunk 6]
```diff
@@ -30,6 +31,7 @@ func CheckSizeLimit() {
 		c := b.Cursor()
 		size := 0
 		for k, v := c.First(); k != nil; k, v = c.Next() {
+			fmt.Println(string(v))
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/boltdb/checker.go]: removed

File: dbservice/dbservice.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	CheckSizeLimit()
+	CheckExpiredData()
+	AggregateScans(output string, currentScan map[string]string, scansPerTicket int, ignoreTheQuantity bool) ([]map[string]string, error)
+	RegisterPlgnInvctn(name string) error
+	EnsureApiKey() error
+	GetApiKey() (string, error)
+	SetDbSizeLimit(limit int)
+}
+
+func ConfigurateDb(id string, dBTestInterval *int, dbMaxSize int) error {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/dbservice.go]: done

File: dbservice/postgresdb/actions.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+func (postgresDb *PostgresDb) MayBeStoreMessage(message []byte, messageKey string, expired *time.Time) (wasStored bool, err error) {
+	db, err := psqlConnect(postgresDb.ConnectUrl)
+	if err != nil {
+		return false, err
+	}
+	defer db.Close()
+
+	if err = initTable(db, dbTableName); err != nil {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/actions.go]: done

File: dbservice/postgresdb/actions.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	currentValue := ""
+	if err = db.Get(&currentValue, fmt.Sprintf("SELECT %s FROM %s WHERE (%s=$1 AND %s=$2)", "messageValue", dbTableName, "id", "messageKey"), postgresDb.Id, messageKey); err != nil {
+		if err != sql.ErrNoRows {
+			return false, err
+		}
+	}
+
+	if currentValue != "" {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/actions.go:24]: when no rows are returned, `db.Get` returns `sql.ErrNoRows` and then checks it:
```
if err != sql.ErrNoRows {
			return false, err
		}
```

File: msgservice/applicationscopeowner_test.go (hunk 4 above)
Comment (DmitriyLewen) [reply msgservice/applicationscopeowner_test.go]: done

File: dbservice/postgresdb/insert.go (hunk 3 above)
Comment (DmitriyLewen) [reply dbservice/postgresdb/insert.go]: removed it

File: dbservice/postgresdb/dbparam.go (hunk 1 above)
Comment (DmitriyLewen) [reply dbservice/postgresdb/dbparam.go]: done

File: dbservice/postgresdb/checker.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+		Date   string `db:"date"`
+		TtlKey string `db:"messagekey"`
+	}
+	if err := db.Select(&scanStructs, fmt.Sprintf("SELECT (key AND ttlkey) FROM %s WHERE %s=$1", dbTableExpiryDates, "id"), postgresDb.Id); err != nil {
+		log.Printf("CheckExpiredData: Can't get %s table: %s", dbTableExpiryDates, err)
+		return
+	}
+
+	max := time.Now().UTC().Format(DateFmt) //remove expired records
+	for _, scanStruct := range scanStructs {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/checker.go]: done

File: dbservice/postgresdb/insert.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
+package postgresdb
+
+import (
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+)
+
+var insert = func(db *sqlx.DB, table, id, columnName2, value2, columnName3, value3 string) error {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/insert.go]: done

File: dbservice/postgresdb/actions_test.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func TestStoreMessage(t *testing.T) {
+	currentValueStoreMessage := ""
+	time := time.Now()
+
+	savedInitTable := initTable
+	initTable = func(db *sqlx.DB, tableName string) error { return nil }
+	savedInsert := insert
+	insert = func(db *sqlx.DB, table, id, columnName2, value2, columnName3, value3 string) error {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/actions_test.go]: done

File: dbservice/postgresdb/init.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/init.go]: done

File: dbservice/postgresdb/actions.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	if err = initTable(db, dbTableName); err != nil {
+		return false, err
+	}
+
+	if err = initTable(db, dbTableExpiryDates); err != nil {
+		return false, err
+	}
+
+	currentValue := ""
+	if err = db.Get(&currentValue, fmt.Sprintf("SELECT %s FROM %s WHERE (%s=$1 AND %s=$2)", "messageValue", dbTableName, "id", "messageKey"), postgresDb.Id, messageKey); err != nil {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/postgresdb/actions.go]: done

File: dbservice/dbservice.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	if *dBTestInterval == 0 {
+		*dBTestInterval = 1
+	}
+
+	if os.Getenv("POSTGRES_URL") != "" {
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply dbservice/dbservice.go]: Done

File: dbservice/postgresdb/actions.go [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+	if err = db.Get(&currentValue, sqlQuery, postgresDb.Id, messageKey); err != nil {
+		if err != sql.ErrNoRows {
+			return false, err
+		}
+	}
+
+	if currentValue != "" {
+		return false, nil
+	} else {
+		if expired != nil {
```
Comment (AndreyLevchenko) [inline dbservice/postgresdb/actions.go:29]: if `expired` is null then
`insertInTableName(db, postgresDb.Id, messageKey, message, expired)` is same to
`insertInTableName(db, postgresDb.Id, messageKey, message, nil)`
so it means if statement is not required

File: dbservice/postgresdb/delete.go [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+var deleteRowsById = func(db *sqlx.DB, table, id string) error {
+	sqlQuery := fmt.Sprintf("DELETE FROM %s WHERE id=$1", table)
+	if _, err := db.Exec(sqlQuery, id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// var deleteRowsByIdAndOutput = func(db *sqlx.DB, id, output string) error {
```
Comment (AndreyLevchenko) [inline dbservice/postgresdb/delete.go:27]: please remove commented code

**Review by AndreyLevchenko — CHANGES_REQUESTED** [review]

File: dbservice/postgresdb/delete.go (hunk 16 above)
Comment (DmitriyLewen) [reply dbservice/postgresdb/delete.go]: done

File: dbservice/postgresdb/actions.go (hunk 15 above)
Comment (DmitriyLewen) [reply dbservice/postgresdb/actions.go]: done

**Comment by AndreyLevchenko** [comment]
merged to `postee-as-api-2` branch

**Pull request closed without merging**